Refactor menu resolvers 9.0

YanaDePauw opened the pull request:
## References
Fixes #3262

## Description
This PR refactors the menu sections that were previously provided in the various menu resolvers. Instead of the previous system where all the sections are added in the same resolver and need to be managed manually through indexes, the new system will split up all the sections over multiple "Providers" which will then be added to the list of providers per menu defined in the `app.menus.ts` file.

```
export const MENUS = buildMenuStructure({  
  [MenuID.PUBLIC]: [  
    CommunityListMenuProvider,  
    BrowseMenuProvider,  
    StatisticsMenuProvider,  
  ],  
  [MenuID.ADMIN]: [  
    NewMenuProvider,  
    EditMenuProvider,  
    ... 
    SystemWideAlertMenuProvider,  
  ],  
  [MenuID.DSO_EDIT]: [  
    DsoOptionMenuProvider.withSubs([  
      SubscribeMenuProvider.onRoute(  
        MenuRoute.SIMPLE_COMMUNITY_PAGE,  
        MenuRoute.SIMPLE_COLLECTION_PAGE,  
      ),  
      ...
      ClaimMenuProvider.onRoute(  
        MenuRoute.SIMPLE_ITEM_PAGE,  
        MenuRoute.FULL_ITEM_PAGE,  
      ),  
    ]),  
  ],  
});
```

This approach makes it way more intuitive to see what kind of sections there are and to add new sections in between already existing ones.

This new version fully removes the old way of adding route based data (e.g. statistics) and replaces it with a new way where you can easily add routes to specified providers and/or retrieve route data in the provider to be used for its sections. It also adds support to separately define a parent sections with various child sections. The latter will be mostly useful for creating sections on the DSO pages where you can now separately define the parent section through a provider and then add child providers for the various different sections.

For example, in the current implementation all the buttons on the DSO pages have been added to a single parent parent section resulting in a single button on the DSO pages with the relevant sections in the dropdown. In case sections need to be moved, the provider can simply be moved to a different location such as before the provider responsible for providing the button (`DsoOptionMenuProvider`).

Finally, since the menus have been moved from the resolvers to the `InitService`, the menus will not delay the page from loading when the menu sections need some time. The menus will simply appear when they are fully loaded. To prevent "blinking" when changing pages, the sections that need to be added/removed are all done at the same time.

## Instructions for Reviewers

The current menus and its sections should be the same except for two exceptions. Therefore, verifying that the behaviour has remained the same for those sections should be sufficient to verify that the refactor was successful.

The two exceptions are the following:

**Exception 1: Statistics section**
The behaviour of the statistics menu has been changed to show a menu on all pages instead of only the home page and the DSO pages.

When a DSO is present somewhere in the route, it will redirect to that DSO's statistics page, otherwise it will redirect to the general statistics page.

For example, when you are on the edit page of an item, the statistics link will still navigate to that item's statistics. When on the "Scripts and processes" page, the statistics link will navigate to the general repository statistics.

**Exception 2:  DSO options wrapper section**
The various buttons on the community, collection and item pages have all been added to a single dropdown menu instead of having separate buttons per section. This results into a cleaner look compared to having a growing list of different buttons being added on those pages.

**Note**: In case people would want to restore the old view, this can be easily done by removing the `DsoOptionMenuProvider.withSubs()`  provider and adding all the menus that were in the .withSubs  list directly to the `[MenuID.DSO_EDIT]` list.

## Documentation
Due to the majority of the changes, some documentation will be provided separately. 

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You do not need to complete this checklist prior creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [ ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR **passes [ESLint](https://eslint.org/)** validation using `npm run lint`
- [x] My PR **doesn't introduce circular dependencies** (verified via `npm run check-circ-deps`)
- [x] My PR **includes [TypeDoc](https://typedoc.org/) comments** for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR **passes all specs/tests and includes new/updated specs or tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **aligns with [Accessibility guidelines](https://wiki.lyrasis.org/display/DSDOC8x/Accessibility)** if it makes changes to the user interface.
- [x] My PR **uses i18n (internationalization) keys** instead of hardcoded English text, to allow for translations.
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

YanaDePauw commented:
## Note: e2e test issues
There are still some e2e tests failing due to these requiring certain ids for certain menus  while most of the menus do not have human readable ids anymore. These are being fixed.

tdonohue commented:
NOTE: This PR has merge conflicts caused by the merger of #3997.  You will need to rebase this PR on the latest `main` and update it to use ["control flow" syntax](https://angular.dev/guide/templates/control-flow) by replacing any `ngIf`, `ngFor`, or `ngSwitch` usages in HTML templates with `@if`, `@for` or `@switch`.  See the control flow docs or the above PR for more details.

tdonohue commented on src/app/shared/menu/providers/access-control.menu.ts:
```diff
@@ -0,0 +1,97 @@
+              link: '/access-control/bulk-access',
+            },
+          },
+          // TODO: enable this menu item once the feature has been implemented
```
I think we should simply remove this commented out code at this point.  It can be added back in later when the feature is added.

YanaDePauw replied:
Removed the TODO.

tdonohue commented on src/app/shared/menu/providers/browse.menu.ts:
```diff
@@ -0,0 +1,69 @@
+    );
+  }
+
+  getSubSections(): Observable<PartialMenuSection[]> {
```
Because this is not a "hardcoded" menu, like many others, we may want to add some code comments here (or at the top of the class) to describe that this menu is getting it's "subsection" options from the current browse definitions defined on the backend.

tdonohue commented on src/app/shared/menu/providers/create-report.menu.ts:
```diff
@@ -0,0 +1,93 @@
+import { AbstractExpandableMenuProvider } from './helper-providers/expandable-menu-provider';
+
+/**
+ * Menu provider to create the report menu sections
```
I'm starting to realize that it would be **very useful** if each of these menus described *where* the menu is found in these comments.

For instance, it's a bit confusing if you look at this list of menus alphabetically, because you'll find a menu on the admin sidebar next to a menu in the top men next to a menu from the Community/Collection/Item page.

Therefore, I'd like to recommend we add better descriptions for each, to make the correct `*.menu.ts` file easier to locate if you want to edit it.

For instance, this one could say:
```
Menu provider to create the "Reports" menu (and subsections) in the admin sidebar
```
The most important part (to me) is to make sure to use the exact same menu name that is used in the User Interface in the comment, so that it's easier to search by name for the correct file to edit.

tdonohue reviewed (changes requested):
@YanaDePauw : I gave this a code review and test today.  Overall, the code looks good, but I have minor comments inline below.

In testing the PR, most everything is working **great**, but I've found a few bugs:

1. The new "Options" drop menu on Community/Collection/Item is failing accessibility tests when it is open (see screenshot).  The failure appears to be caused by the menu items all having `role="menuitem"` but the "Options" menu itself has *no role*.  So, you end up with an accessibility error that says "Certain ARIA roles much be contained by particular parents".

![new-edit-menu](https://github.com/user-attachments/assets/bf7a72f4-5f53-4a18-a722-0cecc75bdf28)

2. In testing the Admin Sidebar, I noticed that several menus are now **missing** because they haven't been refactored as part of this PR.  The two in particular are the "Notifications" menu (added in DSpace 8 for "Quality Assurance" and "Publication Claim") and the "COAR Notify" menu (also added in DSpace 8).  It looks like they have not been refactored.  You can see these missing menus on https://sandbox.dspace.org

3. Something tiny, but noticed when I compared the Admin Menu from this PR and the Admin Menu on https://sandbox.dspace.org.  The "Admin Search" and "Reports" menus have swapped places!  In this PR, the "Reports" menu appears *above* "Admin Search", but on https://sandbox.dspace.org, the "Admin Search" appears above "Reports".

Beyond that, everything else seems to be working well.

tdonohue commented:
@atarix83 : Reminder that this PR could use a code review from you.  Otherwise, I may have to move this forward without your review sometime this week.

tdonohue reviewed (changes requested):
@YanaDePauw : Re-reviewed and re-tested today.  Overall, it's looking better.  But, the accessibility issues still exist.  Here's how I'm able to reproduce them:

1. First, I'm running this PR (obviously) with the UI in production mode (`npm run build:prod; npm run serve:ssr`)
2. I'm using Chrome with the [axe DevTools plugin](https://www.deque.com/axe/devtools/chrome-browser-extension/) installed.  You don't need to purchase anything, just install the free version.
3. Login as an Admin, and go to an Item page.
4. Open the "three dot" menu, so that you can see the "Edit this Item" and "Create new version" menu options.
5. Now, scan the page using axe DevTools by opening up your Chrome DevTools (More Tools -> Developer Tools) and going to the "axe DevTools" tab and click "Full Page Scan"
6. In the results you'll see there will be 3 "Critical" issues found, where all three of them state "Certain ARIA roles must be contained by particular parents").  If you page through all three of them, you'll see the errors are on the menu button, and the two dropdown options.  It will also say `To solve this problem, you need to fix the following: Required ARIA parents role not present: menu, menubar, group`  So, I think the issue is that these `role=menuitem` are not surrounded by a `role=menu` (or similar).

Beyond this accessibility error, everything else looks good to me.  

@atarix83 : Reminder, it'd be good to get your feedback before we merge this PR.

atarix83 reviewed (changes requested):
thanks @YanaDePauw  for this important work

We have reviewed this PR and have some concerns regarding the direction of this implementation. While we agree on the necessity of this significant refactoring, we are worried that the implementation is over-engineered and introduces some boilerplate code. Specifically:

 - There is a provider class for every menu parent entry. This results in a lot of boilerplate code, with all the associated disadvantages.
 - The way the DSO edit can be configured based on routes seems too granular. Generally, if a menu should be applied to the community page, it should also apply to all subpages. While this granularity might seem useful, it could result out in maintenance overhead
 - Managing the new menu implementation could be time-consuming if not well-documented with specific guidelines.

I understand that my review is coming in well beyond the established deadline, so I would like to request only the following changes:

- On DSpace-CRIS, we completely removed the generation of any menu during SSR to reduce the number of requests fetched and to address issues with the extra application cache layer. Could we consider doing the same with DSpace 9?
- The current style of the DSO menu makes it too prominent on the page, with a dropdown that has a primary style. Perhaps we could add the option to highlight some menu items over others within the dropdown menu, similar to what we have on DSpace-CRIS.

![Schermata del 2025-03-27 17-40-20](https://github.com/user-attachments/assets/11f2f32c-c97b-417c-8f81-ecceb0ae293f)

YanaDePauw commented:
@tdonohue I fixed the accessibility issue.

@atarix83 I looked at the feedback and implemented fixes where relevant. 

> There is a provider class for every menu parent entry. This results in a lot of boilerplate code, with all the associated disadvantages.

The boilerplate code is restricted to mostly a class definition, so I disagree that it is a disadvantage. The overview of all the present menus in the app.menus.ts file with each of them just being a small block increases the readability of the code compared to the "blob" of sections where it is hard to see what order they will be rendered in, what sections are child/parent sections,... The code is also more performant in the sense that a page load will no longer be held up by the loading of the menus since they are no longer present in a resolver.  

Note: It is also still possible to put more menus back into a single provider, however we would advise against this because this would reduce the readability of the code again. 

> The way the DSO edit can be configured based on routes seems too granular. Generally, if a menu should be applied to the community page, it should also apply to all subpages. While this granularity might seem useful, it could result out in maintenance overhead

This is a fair point, I reduced the granularity again to correspond to the DSO pages themselves. This is very easy to change since it only requires you to change the constants defined on the routes and in the menu list. Therefore, if in the future more granularity would be needed, it can easily be added again. 

> Managing the new menu implementation could be time-consuming if not well-documented with specific guidelines.

We already have written some documentation which have been added to the wiki [here](https://wiki.lyrasis.org/pages/viewpage.action?pageId=381124782). This will still be further tweaked to contain the latests changes.

>On DSpace-CRIS, we completely removed the generation of any menu during SSR to reduce the number of requests fetched and to address issues with the extra application cache layer. Could we consider doing the same with DSpace 9?

I have added an option to the providers to only render them on browser side. This can be enabled on provider level by setting `renderBrowserOnly` to true. 
By default this is set to false and it has not been enabled anywhere in this PR.

> The current style of the DSO menu makes it too prominent on the page, with a dropdown that has a primary style. Perhaps we could add the option to highlight some menu items over others within the dropdown menu, similar to what we have on DSpace-CRIS.

No new menu styles were added. The expandable dso edit menus have been present in DSpace since the DSO edit menus were added in the original PR but have never been used until now in upstream code. If more styling is needed, this seems to be a separate issue from the refactor. 

Furthermore, to prevent the bloating of the DSO pages with more and more buttons, we decided to use the expandable menu instead. In case the old version (with all separate buttons) should be kept, we can easily fix this by removing the `DsoOptionMenuProvider` wrapper and all menus will be rendered as buttons again.

tdonohue reviewed (approved):
👍 Thanks @YanaDePauw and all.  This looks good to me now.  My feedback has been addressed, and it looks like all menus are present.

I also verified that @atarix83 's minor feedback is addressed.   I think this **definitely** needs to be well documented, so I've added the "needs documentation" flag.  Please add/move all Docs over to our 9.0 Documentation at https://wiki.lyrasis.org/display/DSDOC9x/

Finally, I wanted to note that we need to keep a close eye on this in the Testathon.  I did my best to test all the menus briefly, but it's possible there's things I didn't touch.  This modifies so many menus at once that we'll need to ensure we watch out for any accidental bugs, permission issues or anything else.  Hopefully, they don't come up, but I feel I need to note that because this is being merged in the week before testathon.
